reference/normalize.go

```diff
@@ -0,0 +1,37 @@
+package reference
+
+// Normalizer removes ambiguity in values to ensure
+// parsing is done correctly and without partial valus.
+type Normalizer interface {
+	Canonicalize(string) (string, error)
```

Comment:
The use of the term canonical here and `CanonicalName` is confusing with the idea of the `Canonical` reference. Not sure what would be a more appropriate word.

Also would it make more sense for this interface just to take in a string and do the `Named` parsing. Originally I wanted to leave it as a package function so it could possible access unexported values without trying the docker implementation to the package anymore. However that is not currently the case and the implementation is already accessing unexported types.